Upgrades the TensorFlow backend to TF2.0

josh146 opened the pull request:
**Context:** The existing `'tf'` backend required TensorFlow 1.3, which is an outdated version of TensorFlow not support on latest Python versions.

**Constituent PRs:**

This is a holding branch for PRs related to upgrading the `'tf'` backend to use TensorFlow 2.0. Below is the list of PRs that make up this holding branch:

* #283
* #320  
* #361

ziofil commented:
@josh146 could you fix the conflicts? I think Travis might be on hold because of that

co9olguy commented on tests/integration/test_tf_symbolic.py:
```diff
@@ -565,7 +565,10 @@ def test_eval_true_state_all_fock_probs(self, setup_eng, cutoff, batch_size, tol
             Dgate(-ALPHA) | q[1]
 
         state = eng.run(prog).state
-        probs = state.all_fock_probs(eval=True).flatten()
```
This entire file needs a closer inspection. Most of the tests here were testing the `eval=True/False`. They are still useful as integration tests, and as tests that tensorflow returns tf objects, but the scope of the file has changed significantly now that we don't even support the `eval` keyword behaviour for TF2.0

co9olguy replied:
Also, the `mark.xfail`ed test can be dropped (we no longer support `sessions` or `feed_dict`s). The "reason" message is also a bit misleading

josh146 replied:
@co9olguy: have re-written this tests, making it more of an integration test, as well as removing the redundant tests. I have also added some gradient integration tests, to make sure everything works as it should.

antalszava commented on .github/CHANGELOG.md:
```diff
@@ -2,6 +2,11 @@
 
+* The `"tf"` backend now supports TensorFlow 2.0.
+  [(#283)](https://github.com/XanaduAI/strawberryfields/pull/283)
+  [(#323)](https://github.com/XanaduAI/strawberryfields/pull/323)
```
Just to include all (3 + 1 with this one) PRs?
```suggestion
  [(#320)](https://github.com/XanaduAI/strawberryfields/pull/320)
  [(#323)](https://github.com/XanaduAI/strawberryfields/pull/323)
```

antalszava commented on strawberryfields/backends/tfbackend/circuit.py:
```diff
@@ -50,33 +49,21 @@ class Circuit:
 
     def _make_vac_states(self, cutoff_dim):
         """Make vacuum state tensors for the underlying graph"""
-        with self._graph.as_default():
```
Just curious for my own understanding: how come it was decided to remove having the graph stored?

co9olguy replied:
graphs are no longer used by default in tensorflow2.0

antalszava replied:
Oh got ya, thanks! :blush:

antalszava commented on strawberryfields/backends/tfbackend/circuit.py:
```diff
@@ -399,150 +364,126 @@ def prepare_multimode(self, state, modes, input_state_is_pure=False):
+            return tf.abs(vac_component) ** 2
+        return vac_component
 
     def measure_fock(self, modes, select=None, **kwargs):
```
This function is almost 200 lines long, it might be worth splitting it up for readability. The first ~50 lines are validation checks for example. Out of scope to split up the entire thing, just leaving it (maybe we could open an issue/TODO?).

co9olguy replied:
:grin:

josh146 replied:
Since this is an existing function (i.e., there was no degradation), I saw we leave this for a separate PR.

antalszava commented on strawberryfields/backends/tfbackend/circuit.py:
```diff
@@ -612,161 +551,147 @@ def measure_fock(self, modes, select=None, **kwargs):
-            else:
-                batch_size = 1
-                batch_offset = 0
+        num_reduced_state_modes = len(modes)
```
Following up on the previous comment: this part could now be extracted e.g. as `apply_operation`? And it could return `new_state` to integrate it into the logic. Having a docstring could also help out with further details on the logic.

antalszava commented on strawberryfields/backends/tfbackend/circuit.py:
```diff
@@ -612,161 +551,147 @@ def measure_fock(self, modes, select=None, **kwargs):
 
         # carry out the operation
-        with self.graph.as_default():
-            evaluate_results, session, feed_dict, close_session = ops._check_for_eval(kwargs)
```
Similar question to `graph`: how come `session`, `eval` and `feed_dict` were removed? (For my own understanding)

josh146 replied:
Same reason; having separate sessions, evaluations, and feed dictionaries were part of graph mode and thus not relevant to TensorFlow 2.0

antalszava commented on strawberryfields/backends/tfbackend/circuit.py:
```diff
@@ -803,185 +726,169 @@ def measure_homodyne(self, phi, mode, select=None, **kwargs):
-        with self.graph.as_default():
-            phi = tf.cast(phi, ops.def_type)
-            phi = self._maybe_batch(phi)
+        phi = tf.cast(phi, ops.def_type)
```
Similar thoughts as on `measure_homodyne`: might be worth splitting up this function into using auxiliary functions such that the logical parts are well separated and bugs can hide less easily.

josh146 replied:
Agree!

co9olguy replied:
Worth mentioning that one of the major things people didn't like about TF1.0 was that it was very hard to debug. Previously, splitting up this function would have made little difference for tracking down different bugs. 

Also, some of the functionality here only appears here, and it would possibly be artificial to split into separate functions

antalszava commented on strawberryfields/backends/tfbackend/ops.py:
```diff
@@ -42,7 +42,7 @@
 )
 
-def_type = tf.complex64
+def_type = tf.complex64  # NOTE: what if a user wants higher accuracy?
```
Solution to this could be passing the precision as a keyword argument when choosing the `tf` backend? (Out of scope for now I'd guess)

josh146 replied:
Yep, I would say this counts as a 'new feature' and thus out of scope for now!

thisac commented on strawberryfields/backends/tfbackend/states.py:
```diff
@@ -55,124 +44,72 @@ def __init__(
         r"""
         Computes the trace of the state. May be numerical or symbolic.
 
         Args:
```
Same as my other comment above; should the `Args:` still be there even though it's empty?

josh146 replied:
Removed!

thisac commented on strawberryfields/backends/tfbackend/states.py:
```diff
@@ -55,124 +44,72 @@ def __init__(
         Compute the probabilities of a specific Fock-basis matrix element for the state. May be numerical or symbolic.
 
         Args:
             n (Sequence[int]): the Fock state :math:`\ket{\vec{n}}` that we want to measure the probability of
```
```suggestion
            n (Sequence[int]): the Fock state :math:`\ket{\vec{n}}` that we want to measure the probability of.
```

josh146 replied:
Since this is a sentence fragment, I think it's fine to not have the period

josh146 commented on strawberryfields/backends/tfbackend/backend.py:
```diff
@@ -243,9 +226,6 @@ def state(self, modes=None, **kwargs):
         See :meth:`.BaseBackend.state`.
 
         Keyword Args:
```
I think it should be deleted as well (as its contents was deleted)

josh146 commented on strawberryfields/backends/tfbackend/backend.py:
```diff
@@ -243,9 +226,6 @@ def state(self, modes=None, **kwargs):
         See :meth:`.BaseBackend.state`.
 
         Keyword Args:
```
```suggestion
```

josh146 commented on strawberryfields/backends/tfbackend/backend.py:
```diff
@@ -361,17 +322,8 @@ def measure_homodyne(self, phi, mode, shots=1, select=None, **kwargs):
-            result = np.abs(v - 1) <= tol
-        return result
+            vac_elem = self.circuit.vacuum_element()
+            return np.abs(vac_elem - 1) <= tol
```
How come? 😆

co9olguy replied:
calling `np.something()` on a tensorflow tensor

co9olguy replied:
I'm assuming duck typing saves us here

josh146 replied:
Yep, they've been very clever here (also coupled with the fact that NumPy functions make huge use of duck typing, and very rarely check if the input is an `ndarray`). So as long as `tf.tensor` looks like an iterable (which it does), most NumPy functions will work 🙂

josh146 replied:
Although, there is a slight subtlety here. While `tf.tensor` objects _are_ iterables, `tf.Variable` objects aren't! Which causes a weird edge case, where

```python
>>> x = tf.Variable([0.0, 1.0, 2.0])
>>> list(x) # fails since Variables are not iterable
>>> list(x*1.0) # works since multiplying a variable by a scalar casts it to a pure tensor which is iterable
[<tf.Tensor: id=17, shape=(), dtype=int32, numpy=0>,
 <tf.Tensor: id=21, shape=(), dtype=int32, numpy=1>,
 <tf.Tensor: id=25, shape=(), dtype=int32, numpy=2>]
```

josh146 replied:
So NumPy functions will work great on `tf.constant`, `tf.tensor` objects, but will usually fail directly on `tf.Variable` objects 😆

josh146 commented on strawberryfields/backends/tfbackend/ops.py:
```diff
@@ -246,14 +246,14 @@ def cubic_phase_matrix(gamma, D, hbar, batched=False, method="self_adjoint_eig")
     transpose_list = [1, 0]
     if batched:
         transpose_list = [0, 2, 1]
     if method == "self_adjoint_eig":
```
We should! @ziofil

What was the issue previously @co9olguy, did they have a size limit for matrices?

co9olguy replied:
The previous issue is listed in a code comment directly below this change

josh146 commented on strawberryfields/backends/tfbackend/states.py:
```diff
@@ -55,124 +44,72 @@ def __init__(
     def trace(self, **kwargs):
         r"""
         Computes the trace of the state. May be numerical or symbolic.
 
         Args:
```
```suggestion
```

josh146 commented on tests/backend/test_tf_import.py:
```diff
@@ -35,23 +35,23 @@ class TestBackendImport:
     def test_incorrect_python_version(self, monkeypatch):
         """Test that an exception is raised if the version
-        of Python installed is > 3.6"""
+        of Python installed is 3.5, 3.6 or 3.7"""
```
Actually, we can delete this entire test. TensorFlow 2.0 is available on all currently support python versions.

josh146 commented on tests/backend/test_tf_import.py:
```diff
@@ -71,24 +71,24 @@ class TestFrontendImport:
 
     def test_incorrect_python_version(self, monkeypatch):
         """Test that an exception is raised if the version
         of Python installed is > 3.6"""
```
Updated to delete this test (latest version TF 2.2 supports Python 3.8)

josh146 commented on strawberryfields/backends/tfbackend/states.py:
```diff
@@ -32,20 +32,9 @@ class FockStateTF(BaseFockState):
+    def __init__(self, state_data, num_modes, pure, cutoff_dim, batched=False, mode_names=None):
         # pylint: disable=too-many-arguments
         state_data = tf.convert_to_tensor(
             state_data, name="state_data"
```
done

ziofil commented:
@thisac regarding the empty args/kwargs, I did not remove them to avoid breaking users' code. For instance, if users had kwargs which are now unnecessary they won't have to go change their function calls.

co9olguy commented on tests/integration/test_tf_integration.py:
```diff
@@ -0,0 +1,424 @@
+
+    def test_eval_true_state_all_fock_probs(self, setup_eng, cutoff, batch_size, tol):
+        """Tests whether the Fock-basis probabilities of the state return
+        the a Tensor with the correct value."""
```
```suggestion
        a Tensor with the correct value."""
```

co9olguy commented on tests/integration/test_tf_integration.py:
```diff
@@ -0,0 +1,424 @@
+
+        grad = tape.gradient(res, [a])
+        grad_ex = -a ** 3 * (a ** 2 - 2) * np.exp(-a ** 2)
+        assert np.allclose(grad, grad_ex, atol=tol, rtol=0)
```
Nice changes :ok_hand:  :smile:

nquesada commented on strawberryfields/backends/tfbackend/backend.py:
```diff
@@ -23,7 +23,7 @@
+from .ops import mixed, partial_trace, reorder_modes
 from .states import FockStateTF
 
 
```
The link in line 32 is incorrect, it is pointing to numpy, where it should point to https://www.tensorflow.org/

josh146 replied:
Fixed!

josh146 commented:
>  I noticed one test function names was duplicated, so perhaps one of the tests is actually not being tested

Fixed! Not sure how pytest decides, but the second test with the same name (the more advanced one) was being executed, and the simple test was being skipped!

antalszava commented on .github/CHANGELOG.md:
```diff
@@ -2,11 +2,54 @@
+  [(#373)](https://github.com/XanaduAI/strawberryfields/pull/373)
+  [(#374)](https://github.com/XanaduAI/strawberryfields/pull/374)
+  [(#375)](https://github.com/XanaduAI/strawberryfields/pull/375)
+  [(#377)](https://github.com/XanaduAI/strawberryfields/pull/377)
```
Thanks for this! :)

thisac commented on examples_apps/run_tutorial_machine_learning.py:
```diff
@@ -0,0 +1,367 @@
+    state = result.state
+
+    # Note that all processing, including state-based post-processing,
+
```
```suggestion

```

thisac commented on examples_apps/run_tutorial_machine_learning.py:
```diff
@@ -0,0 +1,367 @@
+
+with tf.GradientTape() as tape:
+    # Here, we map our quantum free parameter `alpha`
+
```
```suggestion

```

co9olguy commented:
> Also, the `examples/quantum_neural_network.py` still seems to be using `tf.Session`. Should/will this be updated?

Yes, that should be updated

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -40,7 +89,8 @@
 This release contains contributions from (in alphabetical order):
 
-Nicolás Quesada
+Tom Bromley, Josh Izaac, Nathan Killoran, Filippo Miatto, Nicolás Quesada, Antal Száva.
```
```suggestion
Tom Bromley, Theodor Isacsson, Josh Izaac, Nathan Killoran, Filippo Miatto, Nicolás Quesada, Antal Száva.
```

josh146 commented:
> > Also, the `examples/quantum_neural_network.py` still seems to be using `tf.Session`. Should/will this be updated?
>
> Yes, that should be updated

Done, updated file is here: https://github.com/XanaduAI/strawberryfields/blob/tf2/examples/quantum_neural_network.py

Thanks for catching that @thisac!

I have also gone through and replaced all remaining occurrences of 'Tensorflow' with 'TensorFlow'.
